File: py/branchaggregate_sdk/features.py
# BranchAggregate SDK feature factory

from branchaggregate_sdk.feature.base_feature import BranchAggregateBaseFeature
from branchaggregate_sdk.feature.debug_feature import BranchAggregateDebugFeature
from branchaggregate_sdk.feature.idempotency_feature import BranchAggregateIdempotencyFeature
from branchaggregate_sdk.feature.metrics_feature import BranchAggregateMetricsFeature
from branchaggregate_sdk.feature.paging_feature import BranchAggregatePagingFeature
from branchaggregate_sdk.feature.ratelimit_feature import BranchAggregateRatelimitFeature
from branchaggregate_sdk.feature.retry_feature import BranchAggregateRetryFeature
from branchaggregate_sdk.feature.test_feature import BranchAggregateTestFeature
from branchaggregate_sdk.feature.timeout_feature import BranchAggregateTimeoutFeature


_FEATURES = {
    "base": lambda: BranchAggregateBaseFeature(),
    "debug": lambda: BranchAggregateDebugFeature(),
    "idempotency": lambda: BranchAggregateIdempotencyFeature(),
    "metrics": lambda: BranchAggregateMetricsFeature(),
    "paging": lambda: BranchAggregatePagingFeature(),
    "ratelimit": lambda: BranchAggregateRatelimitFeature(),
    "retry": lambda: BranchAggregateRetryFeature(),
    "test": lambda: BranchAggregateTestFeature(),
    "timeout": lambda: BranchAggregateTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES

File: php/features.php
<?php
declare(strict_types=1);

// BranchAggregate SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class BranchAggregateFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new BranchAggregateBaseFeature();
            case "debug":
                return new BranchAggregateDebugFeature();
            case "idempotency":
                return new BranchAggregateIdempotencyFeature();
            case "metrics":
                return new BranchAggregateMetricsFeature();
            case "paging":
                return new BranchAggregatePagingFeature();
            case "ratelimit":
                return new BranchAggregateRatelimitFeature();
            case "retry":
                return new BranchAggregateRetryFeature();
            case "test":
                return new BranchAggregateTestFeature();
            case "timeout":
                return new BranchAggregateTimeoutFeature();
            default:
                return new BranchAggregateBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
